Resize for Number of Lines

DominicMDev opened the pull request:
Changed update animation for smoother transition, updating the height in layout subviews to resize for the initial text, and added resizeForNumberOfLines property to allow the textView to scroll after reaching a specified number of lines.

DominicMDev committed:
Changed update animation, updating the height in layout subviews, and added resizeForNumberOfLines property.

Changed update animation for smoother trasition, updating the height in layout subviews to resize for the initial text, and added resizeForNumberOfLines property to allow the textView to scroll after reaching a specified number of lines.

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -86,29 +96,36 @@ open class ResizingTextView: UITextView {
                        delay: 0,
                        options: .layoutSubviews,
-                       animations: { self.layoutIfNeeded() },
+                       animations: { self.heightConstraint.constant = newHeight },
```
Good fix here. I believe this used to not work correctly in an older version of iOS.

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -14,6 +14,16 @@ open class ResizingTextView: UITextView {
     
+    private var maxHeight: CGFloat?
+    
+    public var resizeForNumberOfLines: Int = 0 {
```
Couple things:
1) The name of this variable isn't very clear to me. I'd prefer `maxNumberOfLines` I think.
2) 0 isn't a great default here to disable the behavior. In Objective-C, that would be right, but in Swift, we should just make this an optional `Int?`.

AnthonyMDev replied:
Actually, thinking about it more, if we are doing this, we should allow an explicitly set `maxHeight` as a `CGFloat` also. This should probably be an enum

````
public enum ResizableHeight {
    case height(CGFloat)
    case numberOfLines(Int)
    case infinite
}
````

Then you just have 
````
public var maxResizableHeight: ResizableHeight = .infinite
public var minResizableHeight: ResizableHeight = .numberOfLines(1)
````

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -14,6 +14,16 @@ open class ResizingTextView: UITextView {
+        didSet {
+            guard resizeForNumberOfLines > 0 else { return (maxHeight = nil) }
+            let padding = (textContainer.lineFragmentPadding * 2) + textContainerInset.top + textContainerInset.bottom
+            maxHeight = (font!.lineHeight * CGFloat(resizeForNumberOfLines)) + padding
```
Setting maxHeight here in the `didSet` doesn't account for changes to the `font`, `lineFragmentPadding`, and container insets that might occur AFTER the max lines is set.

We could fix this by doing this calculation in the `didUpdateText(_:)` function. To be a little bit more performant, alternatively, we could listen for changes to the font, padding, and insets on the text container and then cache the max/min height, but I think that's overkill.

I think we should just do this calculation as part of `didUpdateText(_:)`.  I would say to create a private function `calculateHeight(for resizableHeight: ResizableHeight) -> CGFloat?` and call it in `didUpdateText(_:)`. And then you can kill the `maxHeight` variable.

AnthonyMDev replied:
We shouldn't be explicitly unwrapping `font!`. Looks like it can be nil. https://stackoverflow.com/questions/19471467/uitextview-font-is-nil

If the `maxHeight` is being used and the font is `nil`, we should fail to handle the height restriction and just log an error to tell the developer that this is wrong. "Max height cannot be calculated when font is 'nil' for \(self)".

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -86,29 +96,36 @@ open class ResizingTextView: UITextView {
     
+    open override func layoutSubviews() {
+        super.layoutSubviews()
+        didUpdateText(self)
```
This is great for handling the initial size, however, if the user had their text view to have a LARGER initial size than one line (for example in their nib), this could shrink the text view on initial layout.

If we are adding in a maximum size, we should really just add a minimum size also. That would fix this issue by making it explicit that this is going to resize on initial load if needed to be within the max/min range.

Minimum size should default to `.numberOfLines(1)`

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -86,29 +96,36 @@ open class ResizingTextView: UITextView {
+            isScrollEnabled = true
+            showsVerticalScrollIndicator = true
+            return
+        }
```
And then add

````
else if let minHeight = calculateHeight(for minResizableHeight), newHeight < minHeight {
    newHeight = minHeight
}
````

AnthonyMDev reviewed (changes requested):
Thanks a ton for adding this! Great start here, but I don't think that adding a max height without also adding a min height makes sense, due to possible bugs when resizing on initial layout.

I made a bunch of comments on how to improve this and add more functionality. Should be simple to implement from my comments. If you want help implementing it, let me know!

DominicMDev committed:
Created min and max `ResizableHeight`.

Added functionality for specifying a min and max resizable height, where `ResizableHeight` can be: `height(CGFloat)`, `numberOfLines(Int)`, or `infinite`.
Fixed a small miscalculation of the height for number of lines.

DominicMDev committed:
Created min and max `ResizableHeight`

Missed a change for last commit.

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -5,24 +5,34 @@ import UIKit
+    /// Once the `frame`'s height is below this value,
+    /// `showsVerticalScrollIndicator` and `isScrollEnabled` are set back to their previous values.
+    ///
+    /// The defualt value is `ResizableHeight.infinite`
```
Typo on word "default"

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -5,24 +5,34 @@ import UIKit
+    
+    /// The minimum height that the text view should resize to.
+    ///
+    /// The defualt value is `ResizableHeight.numberOfLines(1)`
```
Same as above

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -5,24 +5,34 @@ import UIKit
-            maxHeight = (font!.lineHeight * CGFloat(resizeForNumberOfLines)) + padding
-        }
-    }
+    /// The maximum height that the text view should resize to.
```
Fix documentation, both min and max should have separate documentation.

AnthonyMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -32,6 +42,32 @@ open class ResizingTextView: UITextView {
+    private var initialIsScrollEnabled: Bool = true
+    private var initialShowsVerticalScrollIndicator: Bool = true
+    
+    private var maxHeight: CGFloat? {
```
We don't really need these two helper variables right? I guess I don't mind them but unnecessary.

AnthonyMDev reviewed (approved):
Please fix the documentation issues I mentioned, after that I'll pull this in!

DominicMDev commented on Pod/Classes/ResizingTextView.swift:
```diff
@@ -86,29 +132,58 @@ open class ResizingTextView: UITextView {
+    private func height(forNumberOfLines numberOfLines: Int) -> CGFloat? {
+        guard let font = font else { return nil }
+        let padding = contentInset.top + contentInset.bottom + textContainerInset.top + textContainerInset.bottom
+        return (font.lineHeight * CGFloat(numberOfLines)) + padding
```
> If the `maxHeight` is being used and the font is `nil`, we should fail to handle the height restriction and just log an error to tell the developer that this is wrong. "Max height cannot be calculated when font is 'nil' for (self)".

I currently just have it handling the height as if it is set to `.infinite` when the font is `nil`.
I wasn't sure how we would want the height to be handled in this case. Were you saying you wanted it not to be resized, or were you just saying that there should be some logging involved?

AnthonyMDev replied:
Yeah, using `.infinite` is fine. I was thinking we could add some logging to warn the developer they messed up. Not super important though.

DominicMDev committed:
Added logging

Added logging for failure to calculate height for number of lines when font is nil.

AnthonyMDev reviewed (approved):
Woo!
